src/cryptsetup/cryptsetup.c

```diff
@@ -565,11 +584,96 @@ static char *friendly_disk_name(const char *src, const char *vol) {
         return name_buffer;
 }
 
+static PasswordRegisteredType check_registered_passwords(struct crypt_device *cd) {
+
+        int slots = 0, slot_max;
+        PasswordRegisteredType r = _PASSWORD_REGISTERED_TYPE_INVALID;
+
+        assert(cd);
+
+        if (!streq_ptr(crypt_get_type(cd), CRYPT_LUKS2)) {
+                log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "Non-LUKS2 devices only support passphrase.");
+                return PASSWORD_PASSPHRASE;
+        }
+
+        /* Search all used slots */
+        assert_se((slot_max = crypt_keyslot_max(CRYPT_LUKS2)) > 0);
+        for (int slot = 0; slot < slot_max; slot++) {
+                crypt_keyslot_info status;
+
+                status = crypt_keyslot_status(cd, slot);
+                if (!IN_SET(status, CRYPT_SLOT_ACTIVE, CRYPT_SLOT_ACTIVE_LAST))
+                        continue;
+
+                slots += slot + 1;
+        }
+
+        /* Iterate all LUKS2 tokens and keep track of all their slots */
+        for (int token = 0; token < sym_crypt_token_max(CRYPT_LUKS2); token++) {
+                _cleanup_(json_variant_unrefp) JsonVariant *v = NULL;
+                const char *type;
+                JsonVariant *w, *z;
+                int tk;
+
+                tk = cryptsetup_get_token_as_json(cd, token, NULL, &v);
+                if (IN_SET(tk, -ENOENT, -EINVAL))
+                        continue;
+                if (tk < 0) {
+                        log_warning_errno(tk, "Failed to read JSON token data off disk, ignoring: %m");
+                        continue;
+                }
+
+                w = json_variant_by_key(v, "type");
+                if (!w || !json_variant_is_string(w)) {
+                        log_warning("Token JSON data lacks type field, ignoring.");
+                        continue;
+                }
+
+                type = json_variant_string(w);
+                if (streq(type, "systemd-recovery") || streq(type, "systemd-pkcs11") || streq(type, "systemd-fido2") || streq(type, "systemd-tpm2")) {
```

Comment:
`STR_IN_SET(type, "systemd-recovery", "systemd-pkcs11", …)`

Reply:
Ok, thanks for this hint.